### PR TITLE
[SYCL] Optimize SYCL framework functions with -O0

### PR DESCRIPTION
**Pull request by maksimsab**

### CODE REVIEW

**Comment by maksimsab** [comment]
@elizabethandrews Could you please take a look at FE part?

File: llvm/include/llvm/SYCLLowerIR/IR/PassAdaptors.h
Comment (maksimsab) [inline llvm/include/llvm/SYCLLowerIR/IR/PassAdaptors.h:31]: I added here 2 files in `SYCLLowerIR/IR/PassAdaptors.{h,cpp}` which copies a functionality from `llvm/IR/PassManager.{h,cpp}` of a class `ModuleToFunctionPassAdaptor` but adds a modification that we run a pass on a Function which is marked with `sycl-framework` metadata.

Are we OK with such approach?

File: clang/include/clang/Basic/CodeGenOptions.def
Comment (premanandrao) [inline clang/include/clang/Basic/CodeGenOptions.def:515]:
```suggestion
/// Optimize SYCL Framework functions. These are functions
```

File: clang/test/CodeGenSYCL/sycl-framework-metadata.cpp
Comment (premanandrao) [inline clang/test/CodeGenSYCL/sycl-framework-metadata.cpp:25]: How about something like:

```
namespace {
  namespace sycl {
    void bar4() {}
  }
}
```

File: clang/include/clang/Driver/Options.td
Comment (elizabethandrews) [reply clang/include/clang/Driver/Options.td:2994]: Option used IN conjunction with -O0**

Also, should it be leave instead of left?

File: clang/test/CodeGenSYCL/sycl-framework-metadata.cpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/sycl-framework-metadata.cpp:32]: In this case single_task is not being defined in sycl framework, so it won't have the metadata.  Can you add a test with mock headers to ensure  you are getting behavior you require?

File: clang/test/CodeGenSYCL/sycl-framework-metadata.cpp
Comment (maksimsab) [reply clang/test/CodeGenSYCL/sycl-framework-metadata.cpp:25]: This case is good.

There is a problem that in that case `sycl` namespace becomes ambiguous with the first namespace in the test. I will mitigate it with a separate test file.

File: clang/test/CodeGenSYCL/sycl-framework-metadata1.cpp
Comment (premanandrao) [inline clang/test/CodeGenSYCL/sycl-framework-metadata1.cpp:4]: I think @elizabethandrews meant to also use the single_task interface from the mock header instead of the one defined below for this.

File: clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp
Comment (premanandrao) [inline clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp:19]: Suggest using the interface from the mock header here too.

File: clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp:9]: Can't you add this test to sycl-framework-metadata1.cpp. Why is this second file needed?

Comment (premanandrao) [reply clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp:9]: I think because it leads to an ambiguity with ::sycl otherwise.

Comment (maksimsab) [reply clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp:9]: Yes, this is because of ambiguity.

Comment (maksimsab) [reply clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp:9]: Btw, this is also the reason why I can't use a sycl::kernel_single_task mock from `Inputs/sycl.hpp`

File: clang/test/Driver/sycl-optimize-framework.cpp
Comment (mdtoguchi) [inline clang/test/Driver/sycl-optimize-framework.cpp:5]: I would assume this should also work for Windows?  We should verify that the behavior is correct with `%clang_cl -Od`

File: clang/include/clang/Basic/DiagnosticDriverKinds.td
Comment (mdtoguchi) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:358]: Maybe `err_drv_fsycl_framework_requires_option` and make the option added as a parameter (so you can use the proper option name difference between Linux and MSVC style optimization disabling)

File: clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp
Comment (maksimsab) [reply clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp:19]: It leads to error because of ambiguity of sycl namespace and anonymous namespace.

File: clang/test/Driver/sycl-optimize-framework.cpp
Comment (maksimsab) [reply clang/test/Driver/sycl-optimize-framework.cpp:4]: Added corresponding tests.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td
Comment (maksimsab) [reply clang/include/clang/Basic/DiagnosticDriverKinds.td]: Made the debug option a parameter.

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (elizabethandrews) [inline clang/lib/CodeGen/CodeGenFunction.cpp:1157]: Shouldn't this be done only if flag is passed?

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (elizabethandrews) [inline clang/lib/Frontend/CompilerInvocation.cpp:2077]: Please set using Marshalling infrastructure instead

Comment (maksimsab) [reply clang/lib/Frontend/CompilerInvocation.cpp]: done

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (maksimsab) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: added a check

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (premanandrao) [inline clang/lib/CodeGen/CodeGenFunction.cpp:843]: Can this just be a cast?

**Review by premanandrao — COMMENTED** [review]
One nit/question, but otherwise looks good to me.
@elizabethandrews, had done most of the FE reviews previously.  So please wait for her approval.

**Review by elizabethandrews — APPROVED** [review]
FE changes LGTM

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (maksimsab) [reply clang/lib/CodeGen/CodeGenFunction.cpp:864]: There are plenty of different Derived classes.
https://clang.llvm.org/doxygen/classclang_1_1DeclContext.html

File: llvm/include/llvm/Passes/PassBuilder.h
Comment (maksimsab) [reply llvm/include/llvm/Passes/PassBuilder.h:311]: We could move this part into different file under some *SYCL* directory.

Comment (sarnex) [reply llvm/include/llvm/Passes/PassBuilder.h:311]: My main concern is about the code duplication and decoupling from the upstream unction, My understanding is we probably want to be coupled. @AlexeySachkov Do you have any opinion on this? Thanks

Comment (sarnex) [reply llvm/include/llvm/Passes/PassBuilder.h:311]: Yeah naturally I prefer the second option, but I doubt we can modify the original function, so maybe this is better. I want to see if Alexey has any ideas, he has more experience than me here.

Comment (AlexeySachkov) [reply llvm/include/llvm/Passes/PassBuilder.h:311]: In general, if we really do need some unique pipeline, we can build and maintain it (assuming that we accept the cost of such maintenance), but I doubt that there is a need for such unique pipeline.

My understanding is that by implementing this "framework optimization" feature we simply want to get some acceptable performance in `-O0` mode without sacrificing debuggability of user code much. I suppose that any simple optimization pipeline would do the trick here.

Therefore, my suggestion would be to further explore existing pipeline building functions. For example:

https://github.com/intel/llvm/blob/8c244de9e34140651bdfc2e6ffb4c221a4de1f23/llvm/include/llvm/Passes/PassBuilder.h#L161-L177

This looks like a good candidate to re-use. It is guaranteed to not have any module-level optimizations and likely will produce good enough result for us.

Comment (maksimsab) [reply llvm/include/llvm/Passes/PassBuilder.h]: Significantly simplified optimization pipeline.

File: clang/include/clang/Driver/Options.td
Comment (asudarsa) [inline clang/include/clang/Driver/Options.td:3007]: A better name for this option would be 'fsycl-optimize-sycl-framework-functions'. if there has already been discussion on this, please feel free to ignore my suggestion. Thanks.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td
Comment (sarnex) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:356]:
```suggestion
  "-fsycl-optimize-non-user-code option can be used only in conjunction with %0">;
```
super tiny nitpick

File: clang/include/clang/Driver/Options.td
Comment (sarnex) [inline clang/include/clang/Driver/Options.td:3009]: another super tiny nitpick: i would either make it "with the -O0 flag" or just "with -O0"

File: llvm/lib/Passes/PassBuilderPipelines.cpp
Comment (sarnex) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:1989]: Nice!

File: llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp:19]: Sorry I don't understand what this pass is neded for. I thought in the front end now if you pass -O0 but enable sycl framework O0 optimization, we remove optnone/noinline from SYCL framework functions. Isn't the IR already in the desired state at initial LLVM IR creation time?

File: llvm/include/llvm/SYCLLowerIR/SYCLFrameworkOptimization.h
Comment (sarnex) [inline llvm/include/llvm/SYCLLowerIR/SYCLFrameworkOptimization.h:10]: Sorry I don't funny understand the intent of this pass. Is there a problem where the backend needs noinline/optnone on all functions if any functions use them otherwise it errors?

Also, I assume this happens after all optimization so it doesn't end up actually preventing any optimization of the functions we want to optimize right?

Comment (maksimsab) [reply llvm/include/llvm/SYCLLowerIR/SYCLFrameworkOptimization.h:10]: I was asked to leave IR as "if it was from default debug pipeline".

Yes, it doesn't prevent optimizations of the functions we want to optimize.

**Review by sarnex — APPROVED** [review]
LGTM but might want to get final review from @AlexeySachkov

**Review by smanna12 — APPROVED** [review]

File: clang/lib/CodeGen/CodeGenModule.cpp
Comment (AlexeySachkov) [inline clang/lib/CodeGen/CodeGenModule.cpp:2297]: Why can't we re-use more of existing code and simply set `ShouldAddOptNone` to `false` under this condition?

File: clang/test/CodeGenSYCL/sycl-optimize-non-user-code-func-attrs1.cpp
Comment (AlexeySachkov) [inline clang/test/CodeGenSYCL/sycl-optimize-non-user-code-func-attrs1.cpp:39]: Note that you can capture attribute IDs instead of hardcoding them and give them meaningful names to make test more self-explanatory. Related FileCheck documentation: [here](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks)

File: clang/test/Driver/sycl-optimize-framework.cpp
Comment (AlexeySachkov) [inline clang/test/Driver/sycl-optimize-framework.cpp:15]:
```suggestion
// RUN: %clang_cl -fsycl -Od -fsycl-optimize-non-user-code -### %s 2>&1 | FileCheck %s -check-prefix=OPT-WIN-CHECK
```

File: llvm/lib/Passes/PassBuilderPipelines.cpp
Comment (AlexeySachkov) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:1992]: We are within `buildO0DefaultPipeline` and there is already assertion that optimization level is set to `-O0`: do we really need to duplicate it?

File: llvm/include/llvm/SYCLLowerIR/SYCLFrameworkOptimization.h
Comment (AlexeySachkov) [reply llvm/include/llvm/SYCLLowerIR/SYCLFrameworkOptimization.h:10]: > I was asked to leave IR as "if it was from default debug pipeline".

Can we get a better definition of "it was from default debug pipeline"? Can we get a motivation for that? I just don't quite understand how doing partial optimization of a module co-exists with leaving IR as "if it was from default debug pipeline". To me that extra pass looks weird and it would be great to get more technical background and motivation for its existence.

File: llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp:21]: That doesn't sound right at all. So here we are modifying **all** functions in the module - why?

There are cases when `optnone` can't be added or it will produce invalid LLVM IR:

https://github.com/intel/llvm/blob/1670049aff43dbe529a15704bb66a321696a699f/clang/lib/CodeGen/CodeGenModule.cpp#L2216-L2218

Also, as I said in another comments chain, I don't fully understand the motivation and requirement to have this pass at all and I would prefer not to have and therefore not increase maintenance cost of our compiler.

Note if we decide to keep it, there are no LIT tests for it and that doesn't seem right either.

Comment (asudarsa) [reply llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp:21]: +1.
If we do want to preserve the attributes, we might need a way to save the attributes, remove them and then retrieve them after all optimizations have been applied. Is it a good idea to move the pipeline into sycl-post-link tool where this can be done?

Comment (maksimsab) [reply llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp:21]: sycl-post-link question.
It isn't really a good idea to do this in sycl-post-link tool because of the following reasons:
- Some optimizations require knowledge about target's traits which are provided by [TargetInfo](https://clang.llvm.org/doxygen/classclang_1_1TargetInfo.html) class instance. We don't have a setup of TargetInfo in sycl-post-link as far as I know. So we will be obliged to add a setting up of this stuff in sycl-post-link while it is already done in Clang part.
- There is some infrastructure in clang and llvm which isn't supported in sycl-post-link. For example, you can inspect what passes are being run during compilation with `-mllvm -debug-pass=Structure` (BTW, it is broken with new pass manager, but I believe it is going to be fixed). If we move optimization passes into sycl-post-link then they become invisible for clang user and debugging becomes harder.
- Speaking from architecture point of view, it looks to me that `sycl-post-link` becomes a place where we put any SYCL stuff and it becomes overcomplicated.

About `optnone` correctness issues.
Yes, you are right. `AlwaysInlineAttr` func attr should be taken into account. Motivation is requested in corresponding JIRA tracker.

File: clang/lib/CodeGen/CodeGenModule.cpp
Comment (maksimsab) [reply clang/lib/CodeGen/CodeGenModule.cpp:2297]: There is a problem that in case if we set `ShouldAddOptNone` to `false` then default code path adds `noinline` attribute what is undesirable.

https://github.com/intel/llvm/blob/sycl/clang/lib/CodeGen/CodeGenModule.cpp#L2232-L2236

File: clang/test/CodeGenSYCL/sycl-optimize-non-user-code-func-attrs1.cpp
Comment (maksimsab) [reply clang/test/CodeGenSYCL/sycl-optimize-non-user-code-func-attrs1.cpp]: done

File: clang/lib/CodeGen/CodeGenModule.cpp
Comment (AlexeySachkov) [reply clang/lib/CodeGen/CodeGenModule.cpp:2297]: Got it, thanks

File: llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp
Comment (maksimsab) [reply llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp]: removed this pass.

**Comment by maksimsab** [comment]
Hi @AlexeySachkov !
Do you have any additional thoughts on this one?

**Review by AlexeySachkov — APPROVED** [review]
LGTM

**Comment by bader** [comment]
PR title fix suggestion: add -> Add.

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks

**Review by mdtoguchi — APPROVED** [review]

**Comment by maksimsab** [comment]
@intel/llvm-gatekeepers Can you please merge it?

**Pull request merged**
